db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/ConglomerateDescriptor.java

```diff
@@ -79,6 +80,7 @@
 	private final UUID	uuid;
 	private final UUID	tableID;
 	private final UUID	schemaID;
+	private FirstColumnOfIndexStats firstColumnStats;
```

Comment:
Yes, firstColumnStats is an attribute of the conglomerate.  The AccessPath is updated with the ConglomerateDescriptor of the conglomerate it is currently considering, so we can get it via currentAccessPath->conglomerateDescriptor->firstColumnStats.  Putting it directly into AccessPath would duplicate information and require more code to keep the access path updated with the correct information.  All of the information about the conglomerate currently being considered as an access path is typically collected together in the ConglomerateDescriptor.

I have moved setting of the ConglomerateDescriptor firstColumnStats into StoreCostControllerImpl and synchronized it for threading.  It makes more sense to initialize it here.

Reply:
There was a discussion during the EU code review time between Youssef, Zhen and me regarding this change, some concerns were raised which I'm trying to summarize here for a follow up discussion.
1. This doesn't feel like the right place for FirstColumnOfIndexStats because it doesn't belong semantically in the ConglomerateDescriptor, which should be an immutable description of a conglomerate rather than be updated at runtime with changing statistics. This is mixing different layers and somewhat increasing complexity.
2. Since `ConglomerateDescriptor`s are cached and shared the synchronization for threading could mean some queries get blocked waiting for one another if they get executed concurrently 
3. Are we writing and reading `firstColumnStats` from inside the same `synchronized` block? Otherwise a query could read the stats meant for a different plan.
4. Does `firstColumnStats` always refer to the whole conglomerate? Is it possible it only refers to a part of it if we have a range predicate on the first column?

Since `FirstColumnOfIndexStats` are computed on the fly they might not belong in `StoreCostControllerImpl` either, Zhen suggested introducing a new type of Controller for these runtime computations, but I think a short meeting for a quick discussion and brainstorm would be ideal.

Reply:
@dgomezferro OK, so it seems ConglomerateDescriptor isn't the best place for firstColumnStats.  The alternative pattern is that attributes of a conglomerate are computed in a StoreCostController, and then copied over into a CostEstimate object associated with the access path in ScanCostFunction.generateCost.  At the time this information is copied, the AccessPath is not accessed.  Attributes of the AccessPath seem to reflect more the decisions made by the optimizer in terms of how to access the rows.  Following this pattern, I will move firstColumnStats from ConglomerateDescriptor into StoreCostController and copy it into the CostEstimate.

Reply:
@dgomezferro @ascend1  Regarding the following:

> Since FirstColumnOfIndexStats are computed on the fly they might not belong in StoreCostControllerImpl either, Zhen suggested introducing a new type of Controller for these runtime computations, but I think a short meeting for a quick discussion and brainstorm would be ideal.

I'm thinking that we can add a new dictionary cache to store FirstColumnOfIndexStats associated with each conglomerate, which stays in sync with the partitionStatisticsCache.  This way it doesn't have to be computed on-the-fly each time.  I'm not sure if we want to split it into a different Controller since it is a value derived from tablePartitionStatistics which is also associated with StoreCostController.  What do you think?

Reply:
@dgomezferro Regarding:

> Does firstColumnStats always refer to the whole conglomerate? Is it possible it only refers to a part of it if we have a range predicate on the first column?

It refers to the whole conglomerate.  Currently having a range predicate plus IndexPrefixIteratorMode is not supported.  This would be a useful extension which I thought about implementing.  I don't think it's trivial to do.  We might want to do it as a feature add-on through a separate Jira.